src/httpd_artwork.c

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (C) 2018 Ray/whatdoineed2do <whatdoineed2do@nospam.gmail.com>
+ *
+ * Adapted from httpd_jsonapi.c:
+ * Copyright (C) 2017 Christian Meffert <christian.meffert@googlemail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifdef HAVE_CONFIG_H
+# include <config.h>
+#endif
+
+#include <regex.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "httpd_artwork.h"
+#include "db.h"
+#include "library.h"
+#include "logger.h"
+#include "misc.h"
+#include "player.h"
+#include "artwork.h"
+
+
+static int
+artwork_to_evbuf(struct httpd_request *hreq, struct db_group_info *dbgri, uint32_t *dbmfi_id)
+{
+  int ret = 0;
+  int len = 0;
+  const char* ctype = "invalid";
+  char clen[32];
+  struct evkeyvalq *headers;
+
+  if ( (dbgri == NULL && dbmfi_id == NULL) || (dbgri && dbmfi_id)) 
+    {
+      return -1;
+    }
+
+  if (dbgri)
+    {
+      int  gpid;
+      safe_atoi32(dbgri->id, &gpid);
+      if ( (ret = artwork_get_group(hreq->reply, gpid, 600, 600)) < 0)
+        {
+          DPRINTF(E_LOG, L_WEB, "artwork: failed to retreive for group %d\n", gpid);
+        }
+    }
+  if (dbmfi_id)
+    {
+      if ( (ret = artwork_get_item(hreq->reply, *dbmfi_id, 600, 600)) < 0)
+        {
+          DPRINTF(E_LOG, L_WEB, "artwork: failed to retreive for item %d\n", *dbmfi_id);
+        }
+    }
+
+  len = evbuffer_get_length(hreq->reply);
+  switch (ret)
+    {
+      case ART_FMT_PNG:
+        ctype = "image/png";
+        break;
+
+      case ART_FMT_JPEG:
+        ctype = "image/jpeg";
+        break;
+
+      default:
+        if (len > 0) evbuffer_drain(hreq->reply, len);
+        ret = -1;
+    }
+  DPRINTF(E_DBG, L_WEB, "artwork: via %s  type=%s len=%d\n", (dbgri ? "group" : "item"), ctype, len);
+
+  if (ret)
+    {
+      headers = evhttp_request_get_output_headers(hreq->req);
+      evhttp_remove_header(headers, "Content-Type");
+      evhttp_add_header(headers, "Content-Type", ctype);
+      snprintf(clen, sizeof(clen), "%ld", (long)len);
+      evhttp_add_header(headers, "Content-Length", clen);
+
+      ret = 0;
+    }
+  return ret;
+}
+
+static int
+fetch_artwork(struct httpd_request *hreq, const char *artist_id, const char *album_id)
```

Comment:
What is the purpose of the db lookup? The artwork module will do that for you, so why not just yank the id's straight into artwork_get_xxxx?

Reply:
The web client could ask for artwork for a given `artist` or `album`;  the artwork i/f provides `artwork_get_group()` which requires `group id` but I didn't see anything that could transform a `artist_id` / `album_id` to `group id` for that call hence the db lookups here.

`artwork_get_item()` takes the `media file info` id which again does not appear to pull for the group.

Reply:
Yes, I was thinking the client would use the group id. Is it not available to the client? If not, perhaps it could be?